### ISSUE TITLE
Handling of existing canonical headers

### ISSUE THREAD

**Issue by jeffkaufman** [apache/incubator-pagespeed-mod#1238]
Consider a case where a site has foo.jpg with:

   Link: http://www.jefftk.com/bar.jpg; rel=canonical

And we create xfoo.pagespeed.pagespeed.ic.jykxpX2se3.jpg.  As of 1.10 we would normally put in:

```
Link: <http://www.jefftk.com/foo.jpg>; rel="canonical"
```

Instead, we should put in:

```
Link: <http://www.jefftk.com/bar.jpg>; rel="canonical"
```

Basically, canonical headers should compose.

Reported on the mailing list https://groups.google.com/d/msgid/mod-pagespeed-discuss/CAHRWggSmuCWsoQR%3D5JVQZsy1a583NNZdozMxb8JnA6P9YxEX0A%40mail.gmail.com

**Comment by jeffkaufman** [comment]
Testing current behavior on nginx I get something a little strange:

```
$ headers www.jefftk.com/tempo-meter.jpg?PageSpeed=off
HTTP/1.1 200 OK
Server: nginx/1.8.0
Date: Wed, 06 Jan 2016 15:58:26 GMT
Content-Type: image/jpeg
Content-Length: 60730
Last-Modified: Mon, 04 Jan 2016 20:54:16 GMT
Connection: keep-alive
ETag: "568adbf8-ed3a"
Cache-Control: max-age=600
Link: <http://www.jefftk.com/test-canonical>; rel=canonical
Accept-Ranges: bytes
```

Makes sense: I set up a rel-canonical link to test-canonical.

```
$ headers www.jefftk.com/xtempo-meter.jpg.pagespeed.ic.jykxpX2se3.jpg
HTTP/1.1 200 OK
Server: nginx/1.8.0
Content-Type: image/jpeg
Content-Length: 32261
Connection: keep-alive
Link: <http://www.jefftk.com/tempo-meter.jpg>; rel="canonical"
Date: Wed, 06 Jan 2016 15:57:52 GMT
Expires: Thu, 05 Jan 2017 15:57:52 GMT
Cache-Control: max-age=31536000
ETag: W/"0"
Last-Modified: Wed, 06 Jan 2016 15:57:52 GMT
X-Page-Speed: 1.10.33.2-7600
```

This makes sense as well, given the bug.  We don't currently compose rel=canonical, so we just replace it.

```
$ headers www.jefftk.com/tempo-meter.jpg
HTTP/1.1 200 OK
Server: nginx/1.8.0
Content-Type: image/jpeg
Content-Length: 32261
Connection: keep-alive
ETag: W/"PSA-aj-jykxpX2se3"
Date: Wed, 06 Jan 2016 15:58:23 GMT
Expires: Wed, 06 Jan 2016 16:03:23 GMT
Cache-Control: max-age=300
```

This has no rel=canonical at all, and I think I see why.  When creating `xtempo-meter.jpg.pagespeed.ic.jykxpX2se3.jpg` we added a rel=canonical for `tempo-meter.jpg`, _replacing_ whatever was there already.  And then when serving `tempo-meter.jpg` via ipro we [strip out](https://github.com/pagespeed/mod_pagespeed/commit/61f4e96) any self-referential canonical headers.  So we end up with no header.

At least it's not a duplicate, like @morlovich was worried about?

**Comment by morlovich** [comment]
This might be best to address in the same CL as the flakiness Josh noticed in the integration test for this feature. 

(I was wrong re: duplicates since we don't copy headers wholesale).

**Comment by jeffkaufman** [comment]
@jmarantz

**Comment by morlovich** [comment]
Kind of wondering whether writing a proper parser for link headers is worth it or if approximating is good enough...  Don't think we have anything that quite matches https://tools.ietf.org/html/rfc5988#section-5

**Comment by jmarantz** [comment]
Do we need to parse it?  Or can we just look for
strstr(response_headers->Lookup1("Link"), "canonical") != NULL or something
slightly less crashy & case sensitive?

**Comment by morlovich** [comment]
Yeah, well, that would be roughly what I would call "approximating". It would fail 
on stuff like Link: <www.example.com/canonical/>; rel = next; title="Learning about canonical representations" or such.

Of course, chances are any attempt at proper implementation will have more bugs. Just less known bugs.

**Comment by morlovich** [comment]
Though, maybe we should preserve all the link headers anyway (and only use the guard to not add an extra one)? Not that I consider there being any other than canonical likely.

**Comment by jeffkaufman** [comment]
Preserving them works fine, we don't actually have to do any sort of complicated composition.  Just if there's already a `Link: .*rel.*canonical` then keep that instead of inserting one.

**Comment by jeffkaufman** [comment]
Fixed with 0195e29

Not released yet.
